### PR TITLE
Support for pilot properties and close the job after all batches are processed.

### PR DESCRIPTION
**Pull request by ashitsalesforce**
Support for pilot properties and close the job after all batches are processed.

### CODE REVIEW

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java [hunk 1]
```diff
@@ -69,6 +69,10 @@
     private final LoadRateCalculator rateCalc;
 
     private final boolean updateProgress;
+    
+    private final boolean enablePKChunking = false;
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java:73]: camel casing - enablePKchunking

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java [hunk 2]
```diff
@@ -80,6 +84,29 @@
         } catch (ParameterLoadException e) {
             throw new RuntimeException("Failed to initialize check status interval", e);
         }
+        
+        /*
+         * ======== Start code block to support PK chunking
+         *
+        this.enablePKChunking = ctl.getConfig().getBoolean(Config.ENABLE_BULK_QUERY_PK_CHUNKING);
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java:91]: The value of enablePKChunking was set in line 73, if you are assigning it in this constructor can we omit the assignment in line 73.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java [hunk 3]
```diff
@@ -111,6 +138,19 @@ void createJob(Config cfg) throws AsyncApiException {
...
+            if (this.enablePKChunking) {
+                String startRowParam = "";
+                // startRow parameter of "Sforce-Enable-PKChunking" header has to be a valid
+                // 15 or 18 char ID.
+                if (this.queryChunkStartRow != null 
+                    && (this.queryChunkStartRow.length() == 15 || this.queryChunkStartRow.length() == 18)) {
+                    startRowParam = "; startRow=" + this.queryChunkStartRow;
+                }
+                this.client.addHeader("Sforce-Enable-PKChunking", 
+                        "chunkSize=" + this.queryChunkSize
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java:151]: Can we put line 151 and 152 into one for better readability? Even combining 151 and 512 to line 150 might be better if not too long.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java [hunk 4]
```diff
@@ -169,10 +209,11 @@ boolean hasJob() {
         return this.jobInfo != null;
     }
 
-    void closeJob() throws AsyncApiException {
-        this.jobInfo = this.client.closeJob(getJobId());
+    void awaitCompletionAndCloseJob() throws AsyncApiException {
+        this.jobInfo = this.client.getJobStatus(getJobId());
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java:213]: Does `getJobStatus` still return a `jobInfo`?

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java [hunk 5]
```diff
@@ -71,25 +71,38 @@ protected int executeQuery(String soql) throws AsyncApiException, OperationExcep
...
         }
-        jobUtil.closeJob();
-        final BatchInfo b = jobUtil.getBatches().getBatchInfo()[0];
-        if (b.getState() == BatchStateEnum.Failed) throw new ExtractException("Batch failed: " + b.getStateMessage());
-        this.batch = b;
-        return b.getNumberRecordsProcessed();
+        jobUtil.awaitCompletionAndCloseJob();
+        
+        int recordsProcessed = 0;
+        final BatchInfo[] ba = jobUtil.getBatches().getBatchInfo();
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java:77]: Minor, let's call it something meaning, like`batchInfos`

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java [hunk 6]
```diff
@@ -71,25 +71,38 @@ protected int executeQuery(String soql) throws AsyncApiException, OperationExcep
...
+        for (BatchInfo b : ba) {
```
Comment (jjangsfdc) [inline src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java:78]: Same as here `b` can be `batchInfo`

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java (hunk 1 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java]: Fixed.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java (hunk 2 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java]: This is in the commented out code. That's why I had to set it to false on line 73.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java (hunk 3 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java]: done.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java (hunk 4 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkApiVisitorUtil.java:212]: Yes, it does. the only thing changed is name of the method. Semantics are same as before.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java (hunk 5 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java]: done.

File: src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java (hunk 6 above)
Comment (ashitsalesforce) [reply src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java]: done.

**Pull request merged**